### PR TITLE
feat(portal): Wildcard dns with backwards compatibility

### PR DESCRIPTION
**Pull request by AndrewDryga**
If a new resource is created that will use format not supported by previous client versions we temporarily show a warning:
Screenshot 2024-08-07 at 2 28 57 PM

It will also be excluded from `resources` list for older clients (below 1.2).

### CODE REVIEW

**Comment by AndrewDryga** [comment]
@jamilbk @thomaseizinger I think we need better docs explaining how glob can be used inside the DNS, for example `?` only matches a char, but how do I do 2 chars? (Eg. I want to match `us-east1.foo.com` and `us-east11.foo.com`).

File: elixir/apps/web/lib/web/live/resources/new.ex
Comment (jamilbk) [inline elixir/apps/web/lib/web/live/resources/new.ex]:
```suggestion
                This is an advanced address format. This Resource will be available to Clients v1.2.0 and higher only.
```

File: elixir/apps/web/lib/web/live/resources/new.ex
Comment (jamilbk) [inline elixir/apps/web/lib/web/live/resources/new.ex]:
```suggestion
                class="flex items-center gap-2 text-sm leading-6 text-blue-600 mt-2 w-full"
```

**Comment by thomaseizinger** [comment]
> @jamilbk @thomaseizinger I think we need better docs explaining how glob can be used inside the DNS, for example `?` only matches a char, but how do I do 2 chars? (Eg. I want to match `us-east1.foo.com` and `us-east11.foo.com`).

Agreed! What do you think of a little "testbox" that I suggested? We'd have to re-implement the matching in Elixir (or we compile a shared Rust lib to wasm 🤭).

Also, what do you think of using `*.example.com` instead of `foo.com`? It is after all specifically reserved for docs like this :)

File: elixir/apps/api/test/api/client/channel_test.exs
Comment (thomaseizinger) [inline elixir/apps/api/test/api/client/channel_test.exs:368]: We only support globs at the front currently, not mid domain. Same for question mark.

Comment (thomaseizinger) [reply elixir/apps/api/test/api/client/channel_test.exs:368]: These will need to be removed from the response for < 1.2 clients.

**Review by jamilbk — COMMENTED** [review]
Couple UX suggestions. Thanks for updating docs! I'll leave this to ~~@bmanifold~~ @thomaseizinger for approval.

This will need to be shipped soon after merge because docs go live when hitting `main`.

File: elixir/apps/web/lib/web/live/resources/new.ex
Comment (jamilbk) [reply elixir/apps/web/lib/web/live/resources/new.ex]: This avoids showing a warning during the "happy path".

**Comment by thomaseizinger** [comment]
The compat-test failures suggest that `*.httpbin` gets filtered out incorrectly so the client doesn't know about it and can't connect.

**Comment by thomaseizinger** [comment]
> This will need to be shipped soon after merge because docs go live when hitting `main`.

Should we update the docs separately then?

File: elixir/apps/web/lib/web/live/resources/new.ex
Comment (jamilbk) [reply elixir/apps/web/lib/web/live/resources/new.ex]: There might be a slight usability issue here: if they enter an address that isn't "advanced", behavior is different for Clients less than 1.2 and greater than 1.2. In that case, the help text here is inaccurate. Is that the case?

Comment (thomaseizinger) [reply elixir/apps/web/lib/web/live/resources/new.ex]: The migration from old to new is pretty easy. Maybe we should migrate everything first, ship the new clients but only allow the old format, migrating new entries on the fly before writing them to the DB.

Then we can give users time to upgrade their clients.

File: elixir/apps/api/test/api/client/channel_test.exs
Comment (AndrewDryga) [reply elixir/apps/api/test/api/client/channel_test.exs:368]: @thomaseizinger it's `not in` so the test basically asserts we don't support them.

Comment (thomaseizinger) [reply elixir/apps/api/test/api/client/channel_test.exs:368]: Overlooked the "not" part :)

File: elixir/apps/web/lib/web/live/resources/new.ex
Comment (AndrewDryga) [reply elixir/apps/web/lib/web/live/resources/new.ex]: > There might be a slight usability issue here: if they enter an address that isn't "advanced", behavior is different for Clients less than 1.2 and greater than 1.2. In that case, the help text here is inaccurate. Is that the case?

Not exactly sure how it's different? We take `**.example.com` and map that to `*.example.com` that would be the same, and if you do something 1.1 doesn't support - the client won't see that resource.

Comment (jamilbk) [reply elixir/apps/web/lib/web/live/resources/new.ex]: > if you do something 1.1 doesn't support - the client won't see that resource.

Ah that clears it up. I understand now. Nice!

Comment (jamilbk) [reply elixir/apps/web/lib/web/live/resources/new.ex]: Let me make a quick suggestion to update the helptext

**Comment by AndrewDryga** [comment]
@thomaseizinger should we merge our two PRs into one to make the tests pass? And then we can release it to main.

File: website/src/app/kb/deploy/resources/readme.mdx
Comment (jamilbk) [inline website/src/app/kb/deploy/resources/readme.mdx]: We probably want to merge this to staging and cut v1.2.0 clients before rolling this out to prod, so the time between merge and deploy might be a few days.

I'll add a todo item to uncomment this when we do the aug launch.

```suggestion
  - By default, the pattern will only match exactly the name you enter.
  - To recursively match all subdomains, use a wildcard, such as
    `*.example.com`. This will match `example.com`, `sub2.example.com`, and
    `sub1.sub2.example.com`.
  - To non-recursively match all subdomains, use a question mark, such as
    `?.example.com`. This will match `example.com`, `sub1.example.com`, and
    `sub2.example.com` **but not** `sub1.sub2.example.com`.  
{/*  - By default, the pattern will only match the exact name you enter. */}
{/*  - To match all subdomains recursively, use a double-wildcard, such as */}
{/*    `**.example.com`. This will match `example.com`, `sub.example.com`, and */}
{/*    `sub.sub.example.com`. */}
{/*  - To match all subdomains non-recursively, use a single wildcard, such as */}
{/*    `*.example.com`. This will match `sub.example.com` but not */}
{/*    `sub.sub.example.com`. */}
{/*  - To match a single character, use a question mark, such as */}
{/*    `us-east?.example.com`. This will match `us-east1.example.com` but not */}
{/*    `us-eastXY.example.com`. */}
{/*  - Wildcards can be placed between domain components, e.g., `foo.*.example.com` */}
{/*    will match `foo.bar.example.com` or `foo.**.example.com` will match */}
{/*    `foo.bar.baz.example.com`. */}
```

**Comment by thomaseizinger** [comment]
> @thomaseizinger should we merge our two PRs into one to make the tests pass? And then we can release it to main.

As discussed on Slack, we must be able to merge this on its own, otherwise we are not backwards-compatible.

**Comment by jamilbk** [comment]
What's the status here? Is this blocked on something?

**Comment by thomaseizinger** [comment]
> What's the status here? Is this blocked on something?

I think the last theory we have is that the seeds need to be updated to the new format, otherwise they are being filtered out.

**Review by thomaseizinger — APPROVED** [review]
With the updated seeds, this now passes the integration tests. I can't comment much on the Elixir code. I had a look through the test cases and it seems good!

We should test-drive this on staging extensively before merging the connlib side.

**Review by jamilbk — APPROVED** [review]
Changes LGTM.

When this goes out to staging, we should test with both test clients at 1.2 and released clients at 1.1 to make sure new and existing Resources work.

**Comment by jamilbk** [comment]
Sending this out to staging since I have time to test this today and has been requested by multiple customers.

**Comment by AndrewDryga** [comment]
Thank you @thomaseizinger, sorry I did not handle this before. You did great with Elixir! :)

**Comment by jamilbk** [comment]
@AndrewDryga Hit some issues - looks like we need the compat translation on the Gateway-side too. Otherwise the Gateways fails to decode Resources like `**.com`:

```
    /// One of the stored resources isn't a valid CIDR/DNS.
    #[error("Invalid resource")]
```

More discussion on Slack: https://firezonehq.slack.com/archives/C04HRQTFY0Z/p1723422727069839

**Comment by thomaseizinger** [comment]
@AndrewDryga I opened a PR with a failing test as a starting point here: https://github.com/firezone/firezone/pull/6253

**Pull request merged**
